### PR TITLE
Fix hsv

### PR DESCRIPTION
**Pull request by paulpv**
The existing implementation was for 0 to 255 based H,S,V values.
HSV projects Hue on to a 6 sided hexagon calculated in the 3 RGB channels.
360 / 6 == 60
360 / 3 == 120
The existing code used 43:
255 / 6 == 42 (rounding error to 43)
It also used 85 and 171:
255 / 3 == 85 // between cyan & yellow
255 / 3 * 2 = 170 (rounding error to 171) // between magenta & cyan

So, basically the existing code never worked for lv_color_hsv_t (0-360, 0-100, 0-100).

It also didn't work for 16-bit 565 colors.
You cannot correctly calculate rgbMax - rgbMin if the max and min channels have different color depths (5bit vs 6bit).
There was at least one other reason why the code wouldn't work for non-8bit channels, but I forget what that was now.

Obviously this code uses highly undesirable floating point.
There is perhaps a way to implement this using integer math.
I tried a few times, resorting to #defines, and failed.
I decided that a working floating point impl is still better than a broken integer impl.

I understand if you have a compelling reason to never use floating point.
If so then feel free to insist that this impl be converted to integer math.
I would probably need help with that.

Otherwise, I recommend taking this "fix" and working on an optimization as another PR.

### CODE REVIEW

**Comment by paulpv** [comment]
Most important of all, this code is tested with the following .ino:
https://github.com/paulpv/lv_examples/blob/master/lv_tests/lv_test_misc/lv_test_color/lv_test_color.ino

**Comment by embeddedt** [comment]
I think we should avoid the use of floating point in core LittlevGL code. So far we've managed to do some pretty amazing stuff with just integer math.

Did you try using fixed point? Maybe that would be easier for you to implement.

**Comment by kisvegabor** [comment]
@paulpv 
I agree with @embeddedt. We should avoid using floating-point numbers. If I need extra precision I usually upscale the numbers with 256 or 1024 (`<< 8` or `<< 10`).

**Comment by paulpv** [comment]
Will work on and update PR soon

**Comment by kisvegabor** [comment]
Another issue of HSV functions was spotted by @tgillbe in #1236

**Comment by paulpv** [comment]
I have implemented integer math and will finish up testing and push today

**Comment by paulpv** [comment]
Pushed.
Tested with the following code:
```c
#include <stdio.h>
#include <stdarg.h>
#define LV_LOG_BUF_LEN 1024
char LV_LOG_BUF[LV_LOG_BUF_LEN];

void _lv_log(const char* format, va_list va)
{
    vsnprintf(LV_LOG_BUF, LV_LOG_BUF_LEN, format, va);
    cout << LV_LOG_BUF;
}

void lv_log(const char* format, ...)
{
  va_list va;
  va_start(va, format);
  _lv_log(format, va);
  va_end(va);
}

bool isClose(uint8_t a, uint8_t b, uint8_t d) {
  return LV_MATH_ABS(a - b) <= d;
}

int _testRgbToHsvtoRgb(uint8_t r, uint8_t g, uint8_t b) {
    lv_color_t colorRgbExpected = LV_COLOR_MAKE(r, g, b);
    lv_color_hsv_t colorHsv = lv_color_to_hsv(colorRgbExpected);
    lv_color_t colorRgbActual = lv_color_hsv_to_rgb(colorHsv.h, colorHsv.s, colorHsv.v);
    if (!isClose(colorRgbActual.ch.red, colorRgbExpected.ch.red, 5) || 
        !isClose(colorRgbActual.ch.green, colorRgbExpected.ch.green, 5) || 
        !isClose(colorRgbActual.ch.blue, colorRgbExpected.ch.blue, 5)) {
        lv_log("\ntestRgbToHsvtoRgb: FAIL!\n");
        lv_log("testRgbToHsvtoRgb:              rgb=0x%02X%02X%02X\n", r, g, b);
        lv_log("testRgbToHsvtoRgb: colorRgbExpected=0x%08X\n", colorRgbExpected);
        lv_log("testRgbToHsvtoRgb:         colorHsv={ %d, %d, %d }\n", colorHsv.h, colorHsv.s, colorHsv.v);
        lv_log("testRgbToHsvtoRgb:   colorRgbActual=0x%08X\n", colorRgbActual);
        lv_color32_t colorRgbActual32;
        colorRgbActual32.full = lv_color_to32(colorRgbActual);
        lv_log("testRgbToHsvtoRgb: colorRgbActual32=0x%08X\n", colorRgbActual32);
        return 1;
    }
    return 0;
}

void testRgbToHsvtoRgb() {
    for(uint8_t b = 0; b < 0xFF; b++) {
        for(uint8_t g = 0; g < 0xFF; g++) {
            for(uint8_t r = 0; r < 0xFF; r++) {
                if (_testRgbToHsvtoRgb(r, g, b)) {
                    return;
                }
            }
        }
    }
    lv_log("testRgbToHsvtoRgb: PASS!\n");
}
```

All values from 0xFFFFFF to 0x000000 are tested, and conversion rounding errors are off by at most 5.

File: src/lv_misc/lv_color.c [hunk 1]
```diff
@@ -105,39 +106,66 @@ lv_color_t lv_color_hsv_to_rgb(uint16_t h, uint8_t s, uint8_t v)
...
     }
 
+    // https://en.wikipedia.org/wiki/HSL_and_HSV#Hue_and_chroma
+    int32_t h;
     if(rgbMax == r)
-        hsv.h = 0 + 43 * (g - b) / (rgbMax - rgbMin);
+        h = (((g - b) << 10) / delta) + (g < b ? (6 << 10) : 0); // between yellow & magenta
     else if(rgbMax == g)
-        hsv.h = 85 + 43 * (b - r) / (rgbMax - rgbMin);
+        h = (((b - r) << 10) / delta) + (2 << 10); // between cyan & yellow
```
Comment (kisvegabor) [inline src/lv_misc/lv_color.c:148]: Won't it cause an error if `r > b` as @tgillbe pointed out?

Comment (paulpv) [reply src/lv_misc/lv_color.c:148]: I don't believe so. That would just be a negative number, and the `h < 0` check later bumps it back to positive.

**Comment by paulpv** [comment]
@tgillbe I totally get the improvement of your change over mine. Lets make it happen in the next major version.

**Comment by kisvegabor** [comment]
@embeddedt 
> It might be worthwhile having both versions available (or a way to convert between them), considering that "real" HSV components range from 0..360, 0..100, and 0..100 respectively.

It sounds very reasonable. The question is what to store in `lv_color_hsv_t`? 360/100/100 or 255/255/255. 

I'm thinking about how big problem is that we lose 30% of hue values with 0..255 range. What do you think?

**Comment by paulpv** [comment]
I propose we #define [something along the lines of] "LV_USE_HSV8", where enabling will open up will default lv_color_hsv_t to use a lv_color_hsv8_t struct with 255 based values, and not enabling will use 360/100/100 as it is after this PR.
Then someone can choose which they want, including reduction of hue values, and this won't break any existing users.

**Comment by kisvegabor** [comment]
@paulpv Please see my proposal in [1236-comment](https://github.com/littlevgl/lvgl/pull/1236#issuecomment-546794601).

**Comment by paulpv** [comment]
replied

**Comment by kisvegabor** [comment]
Merged, thank you very much for the contribution!

**Pull request merged**